### DISCUSSION TITLE
Anisotropic elasticity problem

### DISCUSSION THREAD

**Discussion by avtarsinghh1991** [idaholab/moose#19576] [Q&A Modules: General] [answered]
Hello
I am trying to incorporate anisotropic material properties with euler angles for solving simple elasticity problem

Following is the input file.
```
[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Mesh]
  [generated]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 10
    xmax = 1
    ymax = 1
  []
  [pin]
    type = ExtraNodesetGenerator
    input = generated
    new_boundary = pin
    coord = '0 0 0'
  []
[]

[Modules/TensorMechanics/Master]
  [all]
    add_variables = true
    strain = FINITE
    generate_output = 'stress_xx stress_yy vonmises_stress'
  []
[]

#
# Added boundary/loading conditions
# https://mooseframework.inl.gov/modules/tensor_mechanics/tutorials/introduction/step02.html
#
[BCs]
  [left_x]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0
  []
  [bottom_y]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []
  [right_x]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = '0.001*t'
  []
  [pin_x]
    type = DirichletBC
    variable = disp_x
    boundary = pin
    value = 0
  []
  [pin_y]
    type = DirichletBC
    variable = disp_y
    boundary = pin
    value = 0
  []
[]

[Materials]
  [./elasticity_matrix]
    type = ComputeElasticityTensor
    base_name = 'rotation_matrix'
    fill_method = symmetric9
    C_ijkl = '1.684e9 0.176e9 0.176e9 1.684e9 0.176e9 1.684e9 0.754e9 0.754e9 0.754e9'
    # rotation matrix for rotating a vector 30 degrees about the z-axis
    rotation_matrix = '0.8660254 -0.5       0.
                       0.5        0.8660254 0
                       0          0         1'
  [../]
  [./elasticity_euler]
    type = ComputeElasticityTensor
    base_name = 'euler'
    fill_method = symmetric9
    C_ijkl = '1.684e9 0.176e9 0.176e9 1.684e9 0.176e9 1.684e9 0.754e9 0.754e9 0.754e9'
    euler_angle_1 = -30.  # same as above but opposite direction because _transpose_ gets built from these angles
    euler_angle_2 = 0.
    euler_angle_3 = 0.
  [../]

  [stress]
    type = ComputeFiniteStrainElasticStress
  []
[]


# consider all off-diagonal Jacobians for preconditioning
[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  # we chose a direct solver here
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  end_time = 100
  dt = 1
[]

When I am running it. I am getting the following error

*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'elasticity_tensor', requested by 'stress' is not defined on block 0
Material property 'elasticity_tensor', requested by 'stress_face' is not defined on block 0
Material property 'elasticity_tensor', requested by 'stress_neighbor' is not defined on block 0
```

Further, i have my own abaqus mesh file with few grains. How can I read the euler angles and youngs modulus for different grains from the file. Please suggest some reference example.

Please help. Thank you in advance.

**Comment by jiangwen84** [comment, marked as answer]
1. you should not define two `ComputeElasticityTensor`
2. It seems that you want to model polycrystalline materials. If so, please take a look the crystal plasticity models in MOOSE. see https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.html

**Reply by avtarsinghh1991** [reply]
Thanks for your reply. Actually, I am trying to rotate the elastic tensor depending on the crystallographic plane.

C' = RCR^T

in which R is the rotation matrix.

**Reply by jiangwen84** [reply]
You can keep just one `ComputeElasticityTensor` and then remove the `base_name`.

**Reply by avtarsinghh1991** [reply]
Yes, I did the same. I even did a few runs to make sure the orientation is fine. It's working now. Thank you very much.

**Reply by avtarsinghh1991** [reply]
Thanks. Just a follow-up question. I also want to give different Euler angles for different grains. and I want to read those angles from the file. I have my own mesh file with different volume tags for different grains. Can you please suggest some example problems?

Most of the CP model examples are single-crystal only.

**Reply by jiangwen84** [reply]
You can use `ElementPropertyReadFile` with `read_type` option to read in the Euler angles for each block. 

Here is one example

moose/modules/tensor_mechanics/test/tests/crystal_plasticity/user_object_based/prop_block_read.i

**Reply by avtarsinghh1991** [reply]
Thank you. I implemented it and its working.
I am just wondering is that only works with "ComputeElasticityTensorCP" or any material type. Because I can only able to run it for "ComputeElasticityTensorCP".

Below is the input file.

[GlobalParams]
  displacements = 'disp_x disp_y'
[]


[Mesh]
  [generated]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 10
    xmax = 1
    ymax = 1
  []
  [pin]
    type = ExtraNodesetGenerator
    input = generated
    new_boundary = pin
    coord = '0 0 0'
  []
[]


[UserObjects]
  [./euler_angle_read]
    type = ElementPropertyReadFile
    prop_file_name = 'input_file.txt'
    nprop = 3
    read_type = block
    nblock= 1
  [../]
[]

[Modules/TensorMechanics/Master]
  [all]
    add_variables = true
    strain = FINITE
    generate_output = 'stress_xx stress_yy vonmises_stress'
  []
[]


[BCs]
  [left_x]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0
  []
  [bottom_y]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []

  [right_x]
    type = FunctionDirichletBC
    variable = disp_x
    boundary = right
    function = '0.001*t'
  []
  [top_y]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = '0.001*t'
  []


  [pin_x]
    type = DirichletBC
    variable = disp_x
    boundary = pin
    value = 0
  []
  [pin_y]
    type = DirichletBC
    variable = disp_y
    boundary = pin
    value = 0
  []
[]

[Materials]
  [./elasticity_tensor_with_Euler]
    type = ComputeElasticityTensorCP
    fill_method = axisymmetric9
    C_ijkl = '1.984e9 0.196e9 0.176e9 1.684e9 0.176e9 1.584e9 0.754e9 0.754e9 0.754e9'
    read_prop_user_object = euler_angle_read
  [../]
  [stress]
    type = ComputeFiniteStrainElasticStress
  []
[]


[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  end_time = 100
  dt = 1

[]

[Outputs]
  exodus = true
  interval =10
[]

If I am putting material type = "ComputeElasticityTensor" instead of "ComputeElasticityTensorCP". It says "unused parameter 'Materials/elasticity_tensor_with_Euler/read_prop_user_object'"

Or Can I use "ElementPropertyReadFile" for providing other properties to different materials.

Thanks in advance.

**Reply by jiangwen84** [reply]
For elasticity tensors, only `ComputeElasticityTensorCP` is implemented to take Euler angles from `ElementPropertyReadFile` UO.

**Reply by latmarat** [reply]
@jiangwen84, 
when assigning euler angles to grains as "blocks", how do you set the correspondence between moose mesh elements and block ids? 
Also, do you know how moose mesh elements are ordered along x,y,z (i.e., in which order the element labels are increasing)? 
Thanks. 

@asingh-mit, I am also trying to run an elastic simulation with a polycrystal. Would you mind sharing your input file that worked for your elastic polycrystal? 

Thank you all!

**Reply by jiangwen84** [reply]
@latmarat You can either use MOOSE's `SubdomainBoundingBoxGenerator` or use external meshing tools, e.g. Cubit or Gmsh to assign block ids.
